### PR TITLE
Gate diffuse and specular transmission behind shader defs

### PR DESCRIPTION
**Pull request by coreh**
# Objective

- Address #10338

## Solution

- When implementing specular and diffuse transmission, I inadvertently introduced a performance regression. On high-end hardware it is barely noticeable, but **for lower-end hardware it can be pretty brutal**. If I understand it correctly, this is likely due to use of masking by the GPU to implement control flow, which means that you still pay the price for the branches you don't take;
- To avoid that, this PR introduces new shader defs (controlled via `StandardMaterialKey`) that conditionally include the transmission logic, that way the shader code for both types of transmission isn't even sent to the GPU if you're not using them;
- This PR also renames ~~`STANDARDMATERIAL_NORMAL_MAP`~~ to `STANDARD_MATERIAL_NORMAL_MAP` for consistency with the naming convention used elsewhere in the codebase. (Drive-by fix)

---

## Changelog

- Added new shader defs, set when using transmission in the `StandardMaterial`:
  - `STANDARD_MATERIAL_SPECULAR_TRANSMISSION`;
  - `STANDARD_MATERIAL_DIFFUSE_TRANSMISSION`;
  - `STANDARD_MATERIAL_SPECULAR_OR_DIFFUSE_TRANSMISSION`.
- Fixed performance regression caused by the introduction of transmission, by gating transmission shader logic behind the newly introduced shader defs;
- Renamed ~~`STANDARDMATERIAL_NORMAL_MAP`~~ to `STANDARD_MATERIAL_NORMAL_MAP` for consistency;

## Migration Guide

- If you were using `#ifdef STANDARDMATERIAL_NORMAL_MAP` on your shader code, make sure to update the name to `STANDARD_MATERIAL_NORMAL_MAP`; (with an underscore between `STANDARD` and `MATERIAL`)

### CODE REVIEW

File: assets/shaders/array_texture.wgsl [hunk 1]
```diff
@@ -43,7 +43,7 @@ fn fragment(
         double_sided,
         is_front,
 #ifdef VERTEX_TANGENTS
-#ifdef STANDARDMATERIAL_NORMAL_MAP
+#ifdef STANDARD_MATERIAL_NORMAL_MAP
```
Comment (coreh) [inline assets/shaders/array_texture.wgsl:46]: This is a small drive-by fix for consistency with the instances of `STANDARD_MATERIAL` in the codebase.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl [hunk 2]
```diff
@@ -213,6 +213,7 @@ fn apply_pbr_lighting(
         let light_contrib = lighting::point_light(in.world_position.xyz, light_id, roughness, NdotV, in.N, in.V, R, F0, f_ab, diffuse_color);
         direct_light += light_contrib * shadow;
 
+#ifdef STANDARD_MATERIAL_DIFFUSE_TRANSMISSION
         if diffuse_transmission > 0.0 {
```
Comment (mockersf) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:217]: the shader def being present should mean it's > 0.0, could we remove the if from the shader? (same for the others)

**Review by mockersf — APPROVED** [review]
would be nice to get confirmation from someone with the issue on android

**Review by doonv — APPROVED** [review]

File: crates/bevy_pbr/src/render/pbr_functions.wgsl (hunk 2 above)
Comment (coreh) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:217]: Hmm, we could, but with the use of `diffuse_transmission_texture` and `specular_transmission_texture` it's possible for the per-fragment value of `diffuse_transmission` and `specular_transmission` to be 0, even when the shader def is present. In those cases without the if we'd do a bunch or computations or texture reads that would be immediately discarded.

Comment (JMS55) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:217]: I'm not sure how much that really matters. I suspect the overhead of the if statement would be worse than any potential extra work on average. It's not a big deal either way, but it's probably worth removing still.

**Review by JMS55 — APPROVED** [review]

**Comment by coreh** [comment]
> I'm not sure how much that really matters. I suspect the overhead of the if statement would be worse than any potential extra work on average. It's not a big deal either way, but it's probably worth removing still.

Okay, after thinking about it for a bit, I ended up being convinced that it will be a rare enough situation (having a texture to control the transmission levels, and having the majority of it _not_ be transmissive) that's probably not worth it to give the much more common scenario a small penalty every time just for this unlikely benefit. Removed the `if` statements.
